File: app/src/com/njust/helper/library/LibSearchActivity.java
package com.njust.helper.library;

import android.app.SearchManager;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.provider.SearchRecentSuggestions;
import android.support.v4.view.ViewCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;

import com.njust.helper.R;
import com.njust.helper.activity.MyListActivity;
import com.njust.helper.activity.ProgressActivity;
import com.njust.helper.databinding.ItemLibSearchBinding;
import com.njust.helper.localapi.ApiConfiguration;
import com.njust.helper.model.LibSearch;
import com.njust.helper.tools.DataBindingHolder;
import com.zwb.commonlibs.injection.ViewInjection;

import java.util.Collections;
import java.util.List;

public class LibSearchActivity extends ProgressActivity {
    @ViewInjection(R.id.toolbar)
    private Toolbar toolbar;
    @ViewInjection(R.id.searchView)
    private SearchView searchView;
    @ViewInjection(R.id.recyclerView)
    private RecyclerView recyclerView;

    private String search;
    private SearchRecentSuggestions suggestions;
    private LibSearchAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        SearchManager searchManager = (SearchManager) getSystemService(SEARCH_SERVICE);
        searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
        searchView.setQueryRefinementEnabled(true);
    }

    @Override
    protected void prepareViews() {
        adapter = new LibSearchAdapter(this);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
        recyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
    }

    @Override
    protected void setupActionBar() {
        setSupportActionBar(toolbar);
        ViewCompat.setElevation(toolbar, 16f);
        super.setupActionBar();
    }

    @Override
    protected int layoutRes() {
        return R.layout.activity_lib_search;
    }

    public void clear_history(View view) {
        new AlertDialog.Builder(this)
                .setTitle("图书馆")
                .setMessage("您确定清除搜索历史吗？")
                .setPositiveButton("清除", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        getSuggestions().clearHistory();
                    }
                })
                .setNegativeButton(R.string.action_back, null).show();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        String query = intent.getStringExtra(SearchManager.QUERY);
        getSuggestions().saveRecentQuery(query, null);

        search = query;
        attachAsyncTask(new AsyncTask<String, String, List<LibSearch>>() {
            @Override
            protected List<LibSearch> doInBackground(String... params) {
                return ApiConfiguration.getConfiguration("libSearch")
                        .execute(Collections.singletonMap("word", params[0]), LibSearch.class,
                                new ApiConfiguration.ProgressCallback() {
                                    @Override
                                    public void onProgress(String progress) {
                                        publishProgress(progress);
                                    }
                                });
            }

            @Override
            protected void onPostExecute(List<LibSearch> libSearches) {
                if (libSearches != null) {
                    adapter.setData(libSearches);
                }
            }

            @Override
            protected void onProgressUpdate(String... values) {
                showSnack(values[0]);
            }
        }, search);
    }

    private SearchRecentSuggestions getSuggestions() {
        if (suggestions == null) {
            suggestions = new SearchRecentSuggestions(LibSearchActivity.this,
                    SearchSuggestionProvider.AUTHORITY, SearchSuggestionProvider.MODE);
        }
        return suggestions;
    }

    public static class LibSearchAdapter extends MyListActivity.ListRecycleAdapter<LibSearch, ItemLibSearchBinding> {
        private LibSearchActivity activity;

        private LibSearchAdapter(LibSearchActivity activity) {
            this.activity = activity;
        }

        public static void onClick(View view, String id) {
            LibDetailActivity.showLibDetail(view.getContext(), id);
        }

        @Override
        protected int getLayoutRes() {
            return R.layout.item_lib_search;
        }

        @Override
        public void onBindViewHolder(DataBindingHolder<ItemLibSearchBinding> holder, int position) {
            holder.getDataBinding().setLibSearch(getItem(position));
            holder.getDataBinding().setPosition(position);
        }

        @Override
        public void setData(List<LibSearch> data) {
            super.setData(data);
            activity.recyclerView.scrollToPosition(0);
        }
    }

}

File: app/src/main/java/com/njust/helper/library/search/LibSearchActivity.kt
package com.njust.helper.library.search

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.provider.SearchRecentSuggestions
import androidx.activity.compose.setContent
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.njust.helper.BuildConfig
import com.njust.helper.R
import com.njust.helper.library.book.LibDetailActivity
import com.njust.helper.shared.api.LibraryApi
import com.njust.helper.shared.api.ParseErrorException
import com.njust.helper.shared.api.ServerErrorException
import com.zwb.commonlibs.utils.requireSystemService
import kotlinx.coroutines.launch
import java.io.IOException

class LibSearchActivity : AppCompatActivity() {
  private var suggestions: SearchRecentSuggestions? = null
  private val vm = LibSearchViewModel(
    onClickHome = this::finish,
    onClickResultItem = {
      startActivity(LibDetailActivity.buildIntent(this@LibSearchActivity, it.id))
    },
    onClickClearHistory = { clearHistory() },
  ) { searchView ->
    val searchManager = requireSystemService<SearchManager>()
    searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setContent {
      LibSearchScreen(vm = this.vm)
    }
  }

  private fun clearHistory() {
    AlertDialog.Builder(this)
      .setTitle("图书馆")
      .setMessage("您确定清除搜索历史吗？")
      .setPositiveButton("清除") { _, _ -> getSuggestions().clearHistory() }
      .setNegativeButton(R.string.action_back, null).show()
  }

  override fun onNewIntent(intent: Intent) {
    super.onNewIntent(intent)

    val query = intent.getStringExtra(SearchManager.QUERY)
    getSuggestions().saveRecentQuery(query, null)

    onRefresh(query)
  }

  private fun onRefresh(search: String?) {
    if (search == null) {
      return
    }
    vm.isRefreshing = true

    lifecycleScope.launch {
      try {
        val result = LibraryApi.search(search)
        vm.result = result
      } catch (e: Exception) {
        onError(e)
      }
      vm.isRefreshing = false
    }
  }

  private suspend fun onError(throwable: Throwable) {
    when (throwable) {
      is IOException -> showSnack(R.string.message_net_error)
      is ServerErrorException -> showSnack(R.string.message_server_error_lib)
      is ParseErrorException -> showSnack(R.string.message_parse_error)
      else -> if (BuildConfig.DEBUG) {
        throw throwable
      }
    }
  }

  private suspend fun showSnack(resId: Int) {
    vm.snackbarMessageFlow.emit(getString(resId))
  }

  private fun getSuggestions(): SearchRecentSuggestions {
    if (suggestions == null) {
      suggestions = SearchRecentSuggestions(this@LibSearchActivity,
        SearchSuggestionProvider.AUTHORITY, SearchSuggestionProvider.MODE)
    }
    return suggestions!!
  }
}
